### ISSUE TITLE
passwd/unix_chkpwd fails with PAM 1.6.0 and linux-hardened kernel

### ISSUE THREAD

**Issue by hd-gb** [linux-pam/linux-pam#768]
Hello guys,

I am using Arch Linux with the latest linux-hardened kernel from:

* https://archlinux.org/packages/extra/x86_64/linux-hardened/
* https://github.com/anthraxx/linux-hardened

Since pam 1.6.0, I get the following error when running `passwd` as root:

```
[root@archlinux ~]# uname -a 
Linux archlinux 6.7.4-hardened1-1-hardened #1 SMP PREEMPT_DYNAMIC Tue, 13 Feb 2024 19:05:48 +0000 x86_64 GNU/Linux
[root@archlinux ~]# pacman -Q | grep -w pam
pam 1.6.0-3
[root@archlinux ~]# passwd
passwd: Authentication failure
passwd: password unchanged
```

Non-root users are not affected. Downgrading to PAM 1.5.3 solves the issue. Also this does not occur when using the stock linux kernel, so let me know if you think this must be fixed in linux-hardened instead.

I attached the [strace output](https://github.com/linux-pam/linux-pam/files/14279762/strace-passwd.txt) here for further analysis.

Would be grateful for any help :)

**Comment by ldv-alt** [comment]
Could this be something similar to the issue mentioned in https://github.com/linux-pam/linux-pam/issues/747#issuecomment-1908312871?

**Comment by hd-gb** [comment]
Thanks for your fast reply!

I already saw that issue because it seemed related to me at first. However, apparmor is not installed on my system.

Also, I have built a custom PAM version from master branch to see if the patch from [PR 757](https://github.com/linux-pam/linux-pam/pull/757) works for me, but sadly this did not solve my problem either. :(

**Comment by ldv-alt** [comment]
It doesn't have to be apparmor, there could be other restrictions in your hardened system that do not allow `root` to invoke unix_chkpwd helper.

**Comment by stoeckmann** [comment]
I'd like to give more insights to this, but from my understanding of a discussion which I've followed a few weeks ago, this is not a PAM issue, but an issue with rlimits set by shadow's `passwd`, which in turn is not a bug in shadow but it triggers a bug in hardened linux.

A commit within the linux hardened project has been identified to provoke this issue, but it's not fully solved if it can be simply reverted or if another underlying issue has been revealed.

So, in short: It's not PAM. You can notice this also by either using another passwd program which does not set rlimits, maybe busybox, and that you cannot provoke the same issues with tools like `login`, `chsh` other other tools which could be used to modify password entries.

**Comment by hd-gb** [comment]
Can be closed, see https://gitlab.archlinux.org/archlinux/packaging/packages/linux-hardened/-/issues/5

**Comment by thkukuk** [comment]
Thanks for the information.
